### ISSUE TITLE
errors package improvements

### ISSUE THREAD

**Issue by husio** [iov-one/weave#368]
I think that we can improve on how we use errors. Below is a list of some issues I have with current approach and a possible solution to the issue. This is a bucket with errors cleanup ideas :godmode: 

This is a long issue that is intended to be cherry-picked from. Each change that we want to implement should be extracted into its own issue.

## Replace [`Is`](https://godoc.org/github.com/iov-one/weave/errors#Is) function with [`Error.Is`](https://godoc.org/github.com/iov-one/weave/errors#Error) method.

Currently to check an error type, `Is` function must be used.
```go
if _, err := doThings(); errors.Is(errors.ErrNotFound, err) {
    println("not found")
}
```
After making `Is` a method, above example would be
```go
if _, err := doThings(); errors.ErrNotFound.Is(err) {
    println("not found")
}
```
- Current implementation can not support [internal error check](https://github.com/iov-one/weave/blob/master/errors/errors.go#L250).
- Using a method would make it clear what we are comparing with (nicer notation?).
- Comparison with `ErrInternal` would be possible to implement (requires more changes).

## Remove [`Error.New` and `Error.Newf`](https://github.com/iov-one/weave/blob/master/errors/errors.go#L141-L148) methods
`Error.New` method provides the same functionality as `errors.Wrap` function (it is an alias). I think that `Wrap` function provides a better information of what the implementation is doing and what the result is.
Removing those methods will narrow the API from 3 to 1 function for performing the very same operation.

`errors.Wrap` cannot be removed because this is the only way to wrap an `error`. `Error.New` method cannot be used for this.

## Change how errors comparison is implemented
Current implementation is mostly [comparing the ABCI Code value](https://github.com/iov-one/weave/blob/master/errors/errors.go#L257). This is limiting the error testing and prevents from implementing `ErrInternal` comparison.
Instead of the current implementation the following one could be used.
```go
// Is returns true if given error or its cause is the same kind.
// If cause error provides Cause method then a comparison is made with all
// parents as well.
func (kind *Error) Is(err error) bool {
    type causer interface {
        Cause() error
    }
    for {
        if err == kind {
            return true
        }
        if e, ok := err.(causer); ok {
            err = e.Cause()
        } else {
            return false
        }
    }
}
```
Our error instances are global and therefore comparable as instances. Using the above mechanism would also allow to "inherit" errors. Below can be true.
```go
root := Wrap(nil, "root")
child1 := Wrap(root, "child one")
child2 := Wrap(root, "child two")

fmt.Println("child 1 is root", root.Is(child1))
// child 1 is root true

fmt.Println("child 2 is root", root.Is(child2))
// child 2 is root true

fmt.Println("root is child 1", child1.Is(root))
// root is child 1 false

fmt.Println("child 2 is child 1", child1.Is(child2))
// child 2 is child 1 false
```

## Revisit error declarations

We have a lot of error declarations. Many registered in [`errors` package](https://github.com/iov-one/weave/blob/master/errors/errors.go#L13-L75) and more in extensions.

Who is the error code for and what is the expected granularity of the error codes? Can `ErrInvalidInput` cover all cases of an invalid input (ie bad type, string too long, negative number) or should a separate error instance with distinct error code be registered for each case?

Sometimes it is hard to choose which error is apropriate. What is the difference between `ErrInvalidAmount`, `ErrInsufficientAmount` and `ErrInvalidInput`? 
What should be the granularity of an error? A very precise description of each case (ie `ErrOverflow` or `ErrExpired`) or rather general beause wrapping them provides additional information (ie `errors.Wrap(errors.ErrInvalidInput, "age must be a number")`)?

## Revisit the meaning of `ErrInternal`

The design of the tendermint API allows to expose errors to the client. This is does using a numeric error code and a text message.
`weave` takes this design and converts its `error` values into API errors. This is a shortcut that might be simplifing the problem too much.

The original design of current `errors` package was assuming that not all errors are for the clients to be seen. Some `error` values, especially those comming from outside of `weave` can contain an implementaiton details or even secrets. We do not want those to be returned to the client. This is why `ErrInternal` was created. `ErrIntenal` returns a generic "internal error" message without any details.

Our error declarations contains a lot of errors that are implementation details. 
Client is interested in errors like `ErrUnauthorized`, `ErrNotFound` or `ErrInvalidInput`. Those allow to understand why a message was rejected and how to fix the issue.
Client should not receive `ErrOverflow` `ErrHuman` or `ErrInvalidState`. Those are implementation details. They do not help the client.

**Comment by ruseinov** [comment]
FYI we already have Is as both package-level and error-specific method, just using both depending on the context. But it does not support internal errors, that is true.

**Comment by ethanfrey** [comment]
Comments `Replace Is function with Error.Is method`:

This code already exists as a helper.
```
// Is is a proxy helper for global Is to be able to easily instantiate and match error codes
// for example in tests
func (e Error) Is(err error) bool {
	return Is(e.New(""), err)
}
```

Also, we cannot compare two `error` objects with the new format, as that requires an `Error` or `TmError` interface to expose the `Is` method

**Comment by ethanfrey** [comment]
Comment on `Remove Error.New and Error.Newf methods`

These are just helper (convenience) methods over Wrap and I found them quite nice when writing code. As they remove overhead to writing errors, they make me more likely to write better errors.

I would agree that we should add documentation explicitly mentioning them only as helper functions around `Wrap`

**Comment by ethanfrey** [comment]
`Revisit error declarations`

I agree we need to consider what granularity we want, what makes sense as generic errors, what in packages, etc. This was not blocking, but good to agree upon, update [the godoc](https://github.com/iov-one/weave/blob/master/errors/doc.go) accordingly, and then update the packages in question.

Can you please point out some specific examples you believe need to be changed, so we can make more clear guidelines?

**Comment by ethanfrey** [comment]
Comment on `Revisit the meaning of ErrInternal`

I changed this to be non-redacted as the original behavior made it near impossible to debug large classes of errors in tests.

However, if we implement https://github.com/iov-one/weave/issues/363 as proposed, we in my opinion, we can then `Redact` `ErrInternal` as well as `ErrPanic`.

**Comment by ethanfrey** [comment]
Comment on `Change how errors comparison is implemented`

1. I agree with moving from comparing codes to using `Cause` to find a common root registered error.
2. I am unsure if two errors that map down to `ErrInternal` should be considered equal. But *maybe* if these came from `errors.Wrap(errors.ErrInternal, "foo")` rather than `errors.Wrap(randomIoError, "bar")`

**Comment by husio** [comment]
Sorry for clubbing together several topics. Looks like now we need a tree structure for our discussion :sweat_smile: 


> Comments Replace Is function with Error.Is method:  https://github.com/iov-one/weave/issues/368#issuecomment-469422335

I am aware that this method exists. I think that only the method should exist and function `Is` should be removed. In short, I have two reasons to propose that:
1. `Is` function makes it not that obvious what is compared to what. Right now the implementation does not care. It will make a great difference if we will compare instances and not their ABCI codes which I believe we should do.
2. Why to have two ways of doing the very same thing?

> Comment on Remove Error.New and Error.Newf methods https://github.com/iov-one/weave/issues/368#issuecomment-469422746

I think it is important that the code looks simple without a good knowledge of the implementation.
1. `errors.ErrNotFound.New("user")` gives the impression that a **new** error is created. This is not entirely true. Defenietly not true in context of the logic we apply to our errors. Result is still the old error just wrapped. This is not the same as `errors.New("x")` from the standard library.
2. When using weave errors you can use `New` method but when using a value of an `error` type you must use `errors.Wrap`. Why make a distinction here? I think it would be better if all error values were treated the same way.
3. Why provide more than one way to achieve the very same functionality?  


> Revisit error declarations https://github.com/iov-one/weave/issues/368#issuecomment-469424592

I do not have a solution. I have raised this topic because often times I do not know which error to return. This is weird. Because of this I often return whatever error looks good enough at the time.

I think we should either 
1. consider a very thoughtful list of errors that covers all cases and keep as many of them in `errors` package as possible. They will be very granural and there should be almost no need to declare errors in extensions
2. provide a small set of generic errors. For example instead of declaring `ErrWalletNotFound` or `ErrContractNotFound` only the `ErrNotFound` is registered. Use `errors.Wrap(ErNotFound. "wallet")` to create error when wallet is not found.

To point at specific errors, I think [`ErrInvalidMsg`](https://github.com/iov-one/weave/blob/master/errors/errors.go#L30) is terrible. It is an error value returned during the message processing. Of course it is an invalid message (mostly :wink:) !! 
Instead of providing a useless information that the message was invalid an error should contain details what went wrong.

When processing a payment request and there are not enough funds on the account (payment request is for more than the account holds) what should be returned?
- `ErrInvalidMsg` because the message cannot be processed?
- `ErrInvalidInput` because the request is invalid?
- `ErrInvalidAmount` because only the requested amount is invalid?
- `ErrInsufficientAmount` because only the requested amount is greater than the account provides?
- `ErrEmpty` because there are no funds in the wallet?
- `ErrNotFound` because there are no funds in the wallet and therefore the wallet does not exist? 
:see_no_evil: 

> Comment on Revisit the meaning of ErrInternal https://github.com/iov-one/weave/issues/368#issuecomment-469424592
> 
> I changed this to be non-redacted as the original behavior made it near impossible to debug large classes of errors in tests.

Agree that we must think of ease of development as well. Maybe `ErrInternal` was not such a great idea.

I would like to discuss again the idea of exposing all errors as they are to the user. Currently all errors are ending up being returned to the client vial ABCI API. I am not sure if I agree with this. We do not handle errors and instead we just pop them up. Final error is returned to the client to "deal with it". What can a client do?


> Comment on Change how errors comparison is implemented https://github.com/iov-one/weave/issues/368#issuecomment-469426124
>
> 2. I am unsure if two errors that map down to ErrInternal should be considered equal. But maybe if these came from errors.Wrap(errors.ErrInternal, "foo") rather than errors.Wrap(randomIoError, "bar")

This is another thing that I have realized only after it was merged.
When writing tests you cannot check if an internal error was returned. There is no way to compare them :rage3:  So instead of returning `ErrInternal` a different error is returned. This makes the internal error useless.


---

Should I create a separate issue for each topic so that the discussion is easier to follow?

**Comment by ethanfrey** [comment]
> Agree that we must think of ease of development as well. Maybe ErrInternal was not such a great idea.

> I would like to discuss again the idea of exposing all errors as they are to the user. Currently all errors > are ending up being returned to the client vial ABCI API. I am not sure if I agree with this. We do not 
> handle errors and instead we just pop them up. Final error is returned to the client to "deal with it". 
> What can a client do?

Think of the blockchain as a database. All invariants are enforced in the database. The client must have retry logic for errors, or just fail. The abci app just enforces and processes rules. It should not try to interpret the meaning or adjust for the case.

In this case, if send fails as I have not enough, a UI could try sending a lower amount, or give up. Usually we cannot auto-detect the correct action.

**Comment by ethanfrey** [comment]
Most of this has been pulled into #381 and #382 

The rest is generic "think about how specific/generic the errors are" and deserves a number of small issues as we come across concrete issues, like #378 

Can we close this issue now and handle it in the other ones?
